### PR TITLE
Handle Checkout Sessions redirect payment flow in Classic Checkout

### PR DESCRIPTION
**Pull request by diegocurbelo**
Fixes STRIPE-997

## Changes proposed in this Pull Request:

When using redirect-based payment methods (e.g. iDEAL, Bancontact) in the Checkout Sessions flow on classic checkout, the customer was redirected away but returned to the checkout page with no feedback — the order was never created, and no success/error message was shown.

The root cause was that `actions.confirm()` was called **before** the WC order existed, so the `returnUrl` pointed to the checkout page instead of the order-received page.

This PR restructures the Classic Checkout (with Checkout Sessions) flow to work like the current Blocks Checkout:
1. After `loadActions()`, the checkout session ID (now stored during mount) is appended to the form.
2. The form is submitted via AJAX to WooCommerce's checkout endpoint, which creates the order and returns the order-received URL.
3. `actions.confirm()` is then called with the order-received URL as `returnUrl`, so redirect-based payment methods return the customer to the thank-you page.
4. For non-redirect methods (e.g. card), the customer is redirected to the order-received page after confirmation resolves.

Supporting changes:
- `create_checkout_session` PHP endpoint now also returns `session_id` alongside `client_secret`.
- Added `get_checkout_session_status` AJAX endpoint for retrieving a session's status.
- Added `checkoutSessionsGetSessionStatus()` JS API method and corresponding nonce.

## Testing instructions

- Connect to a US Stripe account.
- Enable OCS in the Stripe settings page.
- Enable adaptive pricing under the OCS settings.
- Enable card, iDEAL, and/or Bancontact payment methods.

**Redirect payment method (e.g. iDEAL):**
- As a shopper, add a product to your cart and go to the classic/shortcode checkout page.
- Select iDEAL (or another redirect-based method) and complete checkout.
- Verify you are redirected to the payment provider's page.
- Complete the payment and verify you are returned to the order-received page (not the checkout page).
- As an admin, verify the order is successful and transaction details are present.

**Redirect payment method — cancellation:**
- Repeat the above but cancel the payment at the provider's page.
- Verify you are returned to the order-received page with the order in a pending state.
- Verify the order is eventually marked as failed when the checkout.session.expired webhook fires.

**Non-redirect payment method (card):**
- Pay with a card to verify the existing flow is not broken.

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php [hunk 1]
```diff
@@ -122,7 +157,12 @@ public function create_checkout_session(): void {
 				throw new Exception( __( 'Unable to create Stripe Checkout Session.', 'woocommerce-gateway-stripe' ) );
 			}
 
-			wp_send_json_success( [ 'client_secret' => $checkout_session->client_secret ] );
+			wp_send_json_success(
+				[
+					'client_secret' => $checkout_session->client_secret,
+					'session_id'    => $checkout_session->id,
+				]
+			);
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php:82]: Fixed in [41ea5d4](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/41ea5d4fc6e5ad6d88705dff49d0029100312c49)

File: client/classic/upe/__tests__/payment-processing.test.js [hunk 2]
```diff
@@ -104,8 +116,14 @@ const createMockApi = ( checkoutElements ) => {
...
 		checkoutSessionsCreateSession: jest.fn( () =>
-			Promise.resolve( { data: { client_secret: 'cs_test_abc' } } )
+			Promise.resolve( {
+				data: {
+					client_secret: 'cs_test_abc',
+					session_id: 'cs_test_session_123',
+				},
+			} )
 		),
+		getAjaxUrl: jest.fn( () => '/?wc-ajax=checkout' ),
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply client/classic/upe/__tests__/payment-processing.test.js:149]: Fixed in [bac2c90](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/bac2c90f3e647df28e0b9d956cc1a7047bc99620)

File: client/classic/upe/__tests__/payment-processing.test.js [hunk 3]
```diff
@@ -405,15 +426,21 @@ describe( 'payment-processing', () => {
...
+					confirm: jest.fn().mockResolvedValue( {} ),
 				};
 				const checkoutElements = createMockElements();
 				const api = createMockApi( checkoutElements );
 
+				// Mock the WC checkout AJAX response.
+				mockJQueryAjax.mockResolvedValue( {
+					result: 'success',
+					redirect: orderReceivedUrl,
+				} );
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply client/classic/upe/__tests__/payment-processing.test.js:530]: Fixed in [bac2c90](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/bac2c90f3e647df28e0b9d956cc1a7047bc99620)

File: includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php [hunk 4]
```diff
@@ -15,6 +15,41 @@ class WC_Stripe_Checkout_Sessions_Ajax_Handler {
...
+			if ( ! $is_nonce_valid ) {
+				throw new Exception( __( "We're not able to process this request. Please refresh the page and try again.", 'woocommerce-gateway-stripe' ) );
+			}
+
+			$session_id = isset( $_POST['session_id'] ) ? sanitize_text_field( wp_unslash( $_POST['session_id'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Missing
+			if ( empty( $session_id ) ) {
+				throw new Exception( __( 'Missing checkout session ID.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$checkout_session = WC_Stripe_API::request( [], "checkout/sessions/{$session_id}", 'GET' );
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php]: Fixed in [0007557](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/00075572f45bb45cd6627d75c328667cf6b352ff)

File: includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php [hunk 5]
```diff
@@ -15,6 +15,41 @@ class WC_Stripe_Checkout_Sessions_Ajax_Handler {
...
+
+			if ( is_wp_error( $checkout_session ) || ! is_object( $checkout_session ) || ! empty( $checkout_session->error ) ) {
+				$message = is_object( $checkout_session ) && ! empty( $checkout_session->error->message ) ? $checkout_session->error->message : __( 'Failed to retrieve checkout session.', 'woocommerce-gateway-stripe' );
+				throw new Exception( $message );
+			}
+
+			wp_send_json_success( [ 'status' => $checkout_session->status ?? '' ] );
+		} catch ( Exception $e ) {
+			WC_Stripe_Logger::error( 'Get checkout session status error.', [ 'error_message' => $e->getMessage() ] );
+			wp_send_json_error( [ 'message' => $e->getMessage() ] );
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php]: It's consistent with what `create_checkout_session` returns in case of error; they both use the error message.

File: includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php [hunk 6]
```diff
@@ -15,6 +15,41 @@ class WC_Stripe_Checkout_Sessions_Ajax_Handler {
...
+			if ( ! $is_nonce_valid ) {
+				throw new Exception( __( "We're not able to process this request. Please refresh the page and try again.", 'woocommerce-gateway-stripe' ) );
+			}
+
+			$session_id = isset( $_POST['session_id'] ) ? sanitize_text_field( wp_unslash( $_POST['session_id'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Missing
+			if ( empty( $session_id ) ) {
+				throw new Exception( __( 'Missing checkout session ID.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$checkout_session = WC_Stripe_API::retrieve( "checkout/sessions/{$session_id}" );
```
Comment (diegocurbelo) [inline includes/ajax-handlers/class-wc-stripe-checkout-sessions-ajax-handler.php]: This is debatable, in general we don't want to hardcode what a session_id looks like for Stripe, the plugin should be agnostic to the format.

File: client/api/index.js [hunk 7]
```diff
@@ -687,4 +687,17 @@ export default class WCStripeAPI {
...
 		} );
 	}
+
+	/**
+	 * Retrieves the status of a checkout session.
+	 *
+	 * @param {string} sessionId The Stripe Checkout Session ID.
+	 * @return {Promise} Promise for the request to the server.
+	 */
+	checkoutSessionsGetSessionStatus( sessionId ) {
```
Comment (Mayisha) [inline client/api/index.js]: This seems to be unused. Do we need this and the related Ajax endpoint (`get_checkout_session_status`) in the backend?

Comment (diegocurbelo) [reply client/api/index.js]: Nice catch, it was a leftover from the redirect to checkout on error flow... as we talked in the team call, we are keeping this simple, making it work like the blocks version, and we'll add that redirect "back" in a follow-up PR... removed in [8edf1de](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/8edf1def74f1629633a399e36fe1bb86f6a83f6f)

File: client/classic/upe/__tests__/payment-processing.test.js [hunk 8]
```diff
@@ -573,32 +636,29 @@ describe( 'payment-processing', () => {
...
 
-				expect( mockActions.confirm ).toHaveBeenCalled();
 				expect( stripeUtils.showErrorCheckout ).toHaveBeenCalledWith(
 					'Payment could not be completed. Please try again.'
 				);
 				expect(
 					stripeUtils.appendCheckoutSessionIdToForm
 				).not.toHaveBeenCalled();
-				expect( form.trigger ).not.toHaveBeenCalledWith( 'submit' );
 			} );
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply client/classic/upe/__tests__/payment-processing.test.js]: Fixed in [af17e72](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/af17e72325bcc03b8bce52d9c7de2c4d34e573e2)

File: tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php [hunk 9]
```diff
@@ -25,6 +31,135 @@ public function test_init_hooks(): void {
...
+			$ajax_handler->get_checkout_session_status();
+			$output = ob_get_clean();
+		} catch ( \Exception $e ) {
+			ob_end_clean();
+			throw $e;
+		} finally {
+			remove_filter( 'pre_http_request', $test_request, 10, 3 );
+			Ajax_Test_Helper::remove_hooks();
+			unset( $_POST['session_id'] );
+		}
```
[the remark this answers is not in the document]
Comment (diegocurbelo) [reply tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php]: We deleted the get_checkout_session_status test method and its data provider as part of the "remove unused code" change

**Review by wjrosa — APPROVED** [review]
Looks good to me and works as expected!

Image

Maybe unrelated: but the payment method name is missing above and in the order details page. It should show "Bancontact" here

File: readme.txt [hunk 10]
```diff
@@ -202,5 +202,6 @@ If you get stuck, you can ask for help in the [Plugin Forum](https://wordpress.o
 * Add - Add Ajax endpoint to update line items in a checkout session
 * Tweak - Hide the Adaptive Pricing currency selector from classic checkout when a saved payment method is selected
 * Add - Allow customers to save payment methods during checkout with adaptive pricing
+* Fix - Handle redirect payment flow in classic checkout for Checkout Sessions
```
Comment (Mayisha) [inline readme.txt]: Nit: Can we call it `Add` considering we are adding support for redirect payment in this PR?

File: tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php [hunk 11]
```diff
@@ -385,9 +385,11 @@ public function provide_test_create_checkout_session(): array {
 
 		$checkout_session_missing_secret = (object) [];
 
+		$mocked_session_id = 'cs_test_session_001';
```
Comment (Mayisha) [inline tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php]: `$mocked_session_id` is set twice in this test.

**Review by Mayisha — COMMENTED** [review]
Thanks for working on this @diegocurbelo. The redirect payments are working now and correctly taking me to the order received page. Also the card method is working without any regressions. 👍 

Noting some issues I got during the functional testing.
1. The payment method name in the order received page and order edit page is always `Stripe`. It's not reflecting whether I have paid with iDeal or ACH.

Image

2. If for any reason, the checkout validation fails (for example, select iDeal for payment but keep the name field empty) the cart is getting cleared. I think it could be coming from [this part](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fix/STRIPE-997/handle-redirect-payment-flow-in-classic-checkout/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L1359-L1361) of `process_payment_with_checkout_session` method. This is now an issue because we are calling `/checkout` before `session.confirm()`.
3. After the validation error which clears the cart (mentioned above) if I add the missing info and click the checkout button again (as a user I don't know that the cart got cleared at this point), the checkout page gets stuck on the loading screen.

Image

4. If I simulate a failed payment, it takes me to the order received page which says the order is placed successfully

File: tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php (hunk 11 above)
Comment (diegocurbelo) [reply tests/phpunit/class-wc-stripe-checkout-sessions-ajax-handler-test.php]: Fixed in [f1a3c26](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/f1a3c263d6e61d22702f8a26b52f06258dd497b2)

File: readme.txt (hunk 10 above)
Comment (diegocurbelo) [reply readme.txt]: Fixed in [f1a3c26](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/f1a3c263d6e61d22702f8a26b52f06258dd497b2)

**Review by Mayisha — APPROVED** [review]
The first 3 issues mentioned [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232#pullrequestreview-4077330267) have been fixed. The 4th issue is being handled in STRIPE-1068.

Approving this from my side. 
@wjrosa may we have another look from you on this as I have also added some commits?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 12]
```diff
@@ -1401,11 +1417,6 @@ private function process_payment_with_checkout_session( int $order_id, string $c
 
 		$order->save_meta_data();
 
-		// Remove cart.
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:1404]: Is this removal intentional?

Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:1404]: Yes, that's how the block integration currently works, https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5263 will refactor both Blocks and Classic to use the same flow as PaymentIntents

**Review by wjrosa — CHANGES_REQUESTED** [review]
The "Redirect payment method (e.g. iDEAL)" still works as expected, and now I can see the correct payment method name:

Image

But the cancellation flow ("Redirect payment method — cancellation") redirects me back to the checkout page (I am clicking "Fail payment", maybe that's not the correct test).

Edit: actually let me double-check, I might not have built the plugin files correctly

**Review by wjrosa — APPROVED** [review]
Never mind, my mistake. It is working as expected after the files were built correctly (sorry, too many PRs to review).

But the payment method title is wrong again (maybe unrelated, not sure which PR is supposed to fix it):

Image

**Comment by diegocurbelo** [comment]
> Never mind, my mistake. It is working as expected after the files were built correctly (sorry, too many PRs to review).
> 
> But the payment method title is wrong again (maybe unrelated, not sure which PR is supposed to fix it):
> 

Nice catch @wjrosa, the underlying problem is that the current implementation only relies on webhooks for Checkout Sessions (both blocks and classic)... so if the webhook event takes longer than the redirect the page shows Stripe. 

https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5263 is refactoring the flow to behave like the PaymentIntents flow (sync update on redirect using the params)... but I added an eager title set (in [9bc540f](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5232/commits/9bc540fc9d29c7184d8aa9d7285ebeda0e703c1e)) so the order is created with the selected payment intent.

Screenshot 2026-04-10 at 19 15 08

Can you check again @wjrosa?

**Review by wjrosa — APPROVED** [review]
Now it is correct. Thanks for addressing it so quickly!

Image
